introduce ability to expose engine socket to container

ndeloof opened the pull request:
**- What I did**

introduced ability to opt-in for engine socket to be exposed in container.
this offers an alternative implementation for docker/cli `--use-api-socket`

**- How I did it**

**- How to verify it**

**- Human readable description for the release notes**
```markdown changelog
- Add new mount type `apisocket` to expose the Docker Engine API socket to containers. Containers with an `apisocket` mount automatically receive the `DOCKER_HOST` environment variable pointing to the mounted socket. 
```

**- A picture of a cute animal (not mandatory but encouraged)**

ndeloof committed:
introduce ability to expose engine socket to container

Currently only supported on Linux.

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah reviewed (changes requested):
I'd rather not do this intermediate state at the API yet. We discussed some options for this, but instead of exposing the raw socket (which currently may not be present, depending on the daemon's config), we discussed having an integration at the daemon level that would create a dedicated socket for the container that wants to access the API; this per-container socket would be a proxy for the actual socket, and provide scoped access to the API to allow granular access to the API.

ndeloof commented:
I like the idea to have a new mount type. I can look into this

ndeloof commented on daemon/container.go:
```diff
@@ -124,6 +124,29 @@ func (daemon *Daemon) newContainer(name string, platform ocispec.Platform, confi
+	// Check if any mount is of type apisocket and set DOCKER_HOST environment variable
+	for i, mnt := range hostConfig.Mounts {
+		if mnt.Type == mount.TypeAPISocket {
+			// TODO(ndeloof) engineSocket should create a dedicated socket for id
```
Need to investigate how to register an additional listener (if feasible? Seems this would create a server -> daemon -> sever cycle)

vvoland commented on daemon/container.go:
```diff
@@ -124,6 +124,29 @@ func (daemon *Daemon) newContainer(name string, platform ocispec.Platform, confi
+	// Check if any mount is of type apisocket and set DOCKER_HOST environment variable
+	for i, mnt := range hostConfig.Mounts {
+		if mnt.Type == mount.TypeAPISocket {
+			// TODO(ndeloof) engineSocket should create a dedicated socket for id
```
I think we still need to error out if `Isolation != "none"`?

ndeloof commented on api/types/mount/mount.go:
```diff
@@ -155,3 +158,10 @@ type TmpfsOptions struct {
+// APISocketOptions specifies options specific to mounts of type "apisocket".
+type APISocketOptions struct {
+	Isolation string `json:"isolation"`
+}
```
sure, and also this new mount type to be used by `--use-api-socket` flag

vvoland reviewed (approved):
LGTM

akerouanton reviewed (commented):
I'm on the fence regarding the `isolation` flag that was suggested earlier.

thaJeztah commented on daemon/daemon.go:
```diff
@@ -189,6 +191,21 @@ func (daemon *Daemon) Config() config.Config {
+			return addr, nil
+		}
+	}
+	return "", errors.New("no unix socket configured in daemon hosts")
```
Considering if we should always create a socket for this, but it could be a separate one, e.g. somewhere under `/var/lib/docker`, or some path under the runtime dir. That way we can provide this feature regardless of how the daemon is configured, and it would be "closer" to the per-container socket.

thaJeztah commented on api/types/mount/mount.go:
```diff
@@ -155,3 +158,10 @@ type TmpfsOptions struct {
+	Isolation string `json:"isolation"`
+}
+
+const IsolationNone = "none"
```
> We don't know yet how API restrictions will be surfaced in the API or the UI. We may come up with something that won't fit there. Since its semantics isn't clearly defined, I'd rather not include it and defer that work to a follow-up PR (once we've a better idea of what we want / need).

Yeah, that's the tricky one for this;

- I don't think we want "full access" to be the default; or at least, if we want to provide this feature, we'd rather want a "secure by default" than making full access the default now and later have try users make opt-in to "don't give full access"
- But, yes, we have not designed what options we want to provide and what shape that will have

vvoland commented on api/swagger.yaml:
```diff
@@ -487,6 +489,14 @@ definitions:
+        properties:
+          isolation:
+            type: string
+            enum: ["none"]
```
In future if we have other access options, this is how the Go representation of this look:

```go
type APISocketOptions struct {
    Access string `json:"access"`
    
    // Possibly in future, if we need parameters of other "access" modes:
    PolicyOptions struct {
       PolicyContent string `json:"path"`
       Endpoints []string `json:"endpoints"`
    }  `json:"access-policy"`
    SomeOtherAccessTypeOptions struct{…}
}
```

vvoland reviewed (commented):
Discussed on the maintainer's meeting.

Overall this is looking good!

We had some discussion about:
- Renaming `isolation=none` to `access=unconfined`
  * `unconfined` is more consistent with other options we have (apparmor, seccomp)
- Creating a separate unix socket that would be solely destined for binding into containers.
  * Should be just a matter of adding a new listener, but if there's some issues then not necessarily a blocker

ndeloof commented:
@vvoland pushed https://github.com/moby/moby/pull/51596/changes/2eb76ae8f87a86ba57c34cae8fb15445b174abad

vvoland commented on daemon/daemon_linux.go:
```diff
@@ -237,3 +239,11 @@ func supportsRecursivelyReadOnly(cfg *configStore, runtime string) error {
+func (daemon *Daemon) apiSocket(_ string, opts *dmount.APISocketOptions) (string, error) {
+	if opts != nil && opts.Access != dmount.AccessUnconfined {
+		return "", fmt.Errorf("API socket access %q is not supported", opts.Access)
+	}
```
This will not error out if we don't pass any options at all:


```
curl -X POST -H "Content-Type: application/json"   --unix-socket /var/run/docker.sock   http://localhost/containers/create   -d '{
    "Image": "alpine",
    "Cmd": ["sleep", "3600"],
    "HostConfig": {
       "Mounts": [
        {
          "Type": "apisocket"
        }
      ]
    }
  }'
```

vvoland reviewed (commented):
Could we also have an integration test? 🥺

ndeloof committed:
integration test

Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Debugging

Test is failing, but this looks to be because the defaults are applied
later in the flow, and don't get included in `docker inspect`

    make TEST_FILTER=TestCreateWithAPISocket test-integration
    === FAIL: arm64.integration.container TestCreateWithAPISocket (0.06s)
        create_test.go:868: assertion failed:
            --- inspect.Container.HostConfig.Mounts
            +++ →
              []mount.Mount{
                {
                    Type:        "apisocket",
            - 		Source:      "",
            + 		Source:      "/var/run/docker-api.sock",
            - 		Target:      "",
            + 		Target:      "/var/run/docker.sock",
                    ReadOnly:    false,
                    Consistency: "",
                    ... // 6 identical fields
                },
              }

    time="2025-12-22T13:38:41.360949547Z" level=error msg="linuxParser.parseMountSpec: mount api socket: &{Source: Destination:/var/run/docker.sock RW:true Name: Driver: Type:apisocket Volume:<nil> Mode: Propagation: CopyData:false ID: Spec:{Type:apisocket Source: Target: ReadOnly:false Consistency: BindOptions:<nil> VolumeOptions:<nil> ImageOptions:<nil> TmpfsOptions:<nil> ClusterOptions:<nil> APISocketOptions:0x400096f480} SkipMountpointCreation:false active:0 safePaths:[] Layer:<nil>}"
    time="2025-12-22T13:38:41.360972422Z" level=error msg="daemon.registerMountPoints: mount api socket: &{Source:/var/run/docker-api.sock Destination:/var/run/docker.sock RW:true Name: Driver: Type:apisocket Volume:<nil> Mode: Propagation: CopyData:false ID: Spec:{Type:apisocket Source: Target: ReadOnly:false Consistency: BindOptions:<nil> VolumeOptions:<nil> ImageOptions:<nil> TmpfsOptions:<nil> ClusterOptions:<nil> APISocketOptions:0x400096f480} SkipMountpointCreation:false active:0 safePaths:[] Layer:<nil>}"

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

ndeloof commented on daemon/volumes.go:
```diff
@@ -293,6 +293,7 @@ func (daemon *Daemon) registerMountPoints(ctr *container.Container, defaultReadO
 				return err
 			}
 			mp.Source = socket
+			log.L.Errorf("daemon.registerMountPoints: mount api socket: %+v", mp)
```
I don't get why this happens. `registerMountPoints` is called before `daemon.register(ctx, ctr)` so I expected container's metadata to be set with updated socket mount details

thaJeztah committed:
WIP: quick fixes

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
